Mathlib/Topology/ContinuousFunction/FunctionalCalculus.lean

```diff
@@ -0,0 +1,676 @@
+related to continuity are dispatched by (a small wrapper around) `fun_prop`. As for goals involving
+the predicate `p`, it should be noted that these will only ever be of the form `IsStarNormal a`,
+`IsSelfAdjoint a` or `0 ≤ a`. For the moment we provide a rudimentary tactic to deal with these
+goals, but it can be modified to become more sophisticated as the need arises.
+-/
+
+section Basic
+
+/-- A star `R`-algebra `A` has a *continuous functional calculus* for elements satisfying the
+property `p : A → Prop` if
+
++ for every such element `a : A` there is a star algebra homomorphism
+  `cfcSpec : C(spectrum R a, R) →⋆ₐ[R] A` sending the (restriction of) the identity map to `a`.
++ `cfcSpec` is a closed embedding for which the spectrum of the image of function `f` is its range.
++ `cfcSpec` preserves the property `p`.
+
+The property `p` is marked as an `outParam` so that the user need not specify it. In practice,
+
++ for `R := ℂ`, we choose `p := IsStarNormal`,
++ for `R := ℝ`, we choose `p := IsSelfAdjoint`,
++ for `R := ℝ≥0`, we choose `p := (0 ≤ ·)`.
+-/
+class ContinuousFunctionalCalculus (R : Type*) {A : Type*} (p : outParam (A → Prop))
+    [CommSemiring R] [StarRing R] [MetricSpace R] [TopologicalSemiring R] [ContinuousStar R]
+    [Ring A] [StarRing A] [TopologicalSpace A] [Algebra R A] where
+  /-- The star algebra homomorphisms of the continuous functional calculus. -/
+  toStarAlgHom {a} (ha : p a) : C(spectrum R a, R) →⋆ₐ[R] A
+  /-- The continuous functional calculus is a closed embedding. -/
+  hom_closedEmbedding {a} (ha : p a) : ClosedEmbedding <| toStarAlgHom ha
+  /-- The continuous functional calculus extends the polynomial functional calculus. -/
+  hom_id {a} (ha : p a) : toStarAlgHom ha ((ContinuousMap.id R).restrict <| spectrum R a) = a
+  /-- The continuous functional calculus satisfies the spectral mapping property. -/
+  hom_map_spectrum {a} (ha : p a) : ∀ f, spectrum R (toStarAlgHom ha f) = Set.range f
+  /-- The continuous functional calculus preserves the predicate. -/
+  predicate_hom {a} (ha : p a) : ∀ f, p (toStarAlgHom ha f)
+
+/-- A class guaranteeing that the continuous functional calculus is uniquely determined by the
+properties that it is a continuous star algebra homomorphism mapping the (restriction of) the
+identity to `a`. This is the necessary tool used to establish `cfcSpec_comp` and the more common
+variant `cfc_comp`.
+
+This class will have instances in each of the common cases `ℂ`, `ℝ` and `ℝ≥0` as a consequence of
+the Stone-Weierstrass theorem. -/
+class UniqueContinuousFunctionalCalculus (R A : Type*) [CommSemiring R] [StarRing R]
+    [MetricSpace R] [TopologicalSemiring R] [ContinuousStar R] [Ring A] [StarRing A]
+    [TopologicalSpace A] [Algebra R A] : Prop where
+  eq_of_continuous_of_map_id (a : A) (φ ψ : C(spectrum R a, R) →⋆ₐ[R] A)
+    (hφ : Continuous φ) (hψ : Continuous ψ)
+    (h : φ (.restrict (spectrum R a) <| .id R) = ψ (.restrict (spectrum R a) <| .id R)) :
+    φ = ψ
```

Comment:
Something that's worth mentioning is why you need two classes here instead of a single one.